idelib/transforms.py

```diff
@@ -108,16 +120,113 @@ def isValid(self, session=None, noBivariates=False):
         except AttributeError:
             return False
 
+
+    def inplace(self, values, y=None, timestamp=None, session=None, noBivariates=False, out=None):
+        """ In-place Transform call for the Transform base class.
+            All subclasses are required to implement this method, so it will
+            raise an error if called.
+        """
+        raise NotImplementedError()
+
+
+    @property
+    def useMean(self):
+        """ Property to determine if a transform will use the mean of the
+            secondary channel.  The base class and all univariate polynomials
+            do not have a secondary channel, so they will always return the
+            default `True` and generate a warning.
+        """
+        warnings.warn(UserWarning('{} does not support useMean'.format(type(self))))
+        return True
+
+
+    @useMean.setter
+    def useMean(self, value):
+        warnings.warn(UserWarning('{} does not support useMean'.format(type(self))))
+
+
+    def addWatcher(self, watcher):
+        """ Adds `watcher` to the list (a `WeakSet`) of watchers.  The watchers
+            are other polynomials, such as a `CombinedPoly`, which reference
+            this polynomial.  This is used to propogate changes to all
+            polynomials which reference this one.
+        """
+        self._watchers.add(watcher)
+
     
     #===========================================================================
     # 
     #===========================================================================
-    
+
     @classmethod
     def null(cls, *args, **kwargs):
         return args[0]
 
 
+class ComplexTransform(Transform):
+    """ An mixin class for arbitrarily complex functions that don't follow a set form.
+
+        Instances are function-like objects that take one argument: a sensor
+        reading.
+    """
+
+    def inplace(self, values, y=None, timestamp=None, session=None, noBivariates=False, out=None):
+        """ In-place transform for the `ComplexTransform` transform.  These functions
+            can't be easily reduced and combined like polynomials, so this instead
+            wraps the `function` method.
+        """
+
+        scalar = np.isscalar(values)
+
+        if scalar:
+            values = float(values)
+        elif out is None:
+            out = np.zeros_like(values, dtype=np.float64)
+
+        if len(self._variables) == 1:
+
+            if scalar:
+                out = self.function(values)
+            else:
+                out[:] = self.function(values)
```

Comment:
I might change things to be more aware of 0d arrays, but right now they're definitely an edge case.  This is for a situation where the value passed in is just a regular int/float value, and the user is expecting the same to be returned.

Reply:
My comment was less "we should support 0-D arrays" and more "why not represent scalars as 0-D arrays?" They should behave the same way in any numpy operations, and it'd cut down on special-case logic.

Reply:
Just did some quick testing, since that's an interesting idea.  Unfortunately, the following code is invalid:

```python
a = np.array(1)
a[:] = 2
```

Basically, you can't index into a 0-D array.

I'm going to play around with this a bit more and see if I can make it work a different way, since it looks like type promotion will always turn it back into an int/float.

Reply:
This looks like a cleaner solution, but especially given the age of this PR I'd prefer to document this in an issue and add it in later.

Reply:
yeah `a[:] = 2` will fail but `a[...] = 2` should work fine.

But I agree this should go in an issue, and not hold up this PR.